File: java/PracticaPDOO/src/main/java/deepspace/Weapon.java
/**
 *
 * @author Alejandro Villanueva Prados
 */
package deepspace;

public class Weapon {
    private String name;
    private WeaponType type;
    private int uses;
    
    Weapon(String n, WeaponType t, int u) {
        name = n;
        type = t;
        uses = u;
    }
    
    Weapon(Weapon otro) {
        name = otro.name;
        type = otro.type;
        uses = otro.uses;
    }
    
    WeaponType getType() {
        return type;
    }
    
    int getUses() {
        return uses;
    }
    
    float power() {
        return type.getPower();
    }
    
    float useIt() {
        if (getUses() > 0)
            return power();
        else 
            return 1.0f;
    }
    
    @Override
    public String toString() {
        return "Weapon[" + "@" + System.identityHashCode(this) + ", " + name + ", " + getType() + ", " + getUses() + "]";
    }
}

File: ruby/Weapon.rb
require_relative "WeaponToUI"

module Deepspace
  class Weapon
    def initialize(n, wt, u)
      @name = n
      @type = wt
      @uses = u
    end

    attr_reader :type, :uses

    def self.newCopy(otro)
      otro.clone
    end

    def getType
      @type
    end

    def getUses
      @uses
    end

    def power
      @type.getPower
    end

    def useIt
      if @uses > 0
        @uses -= 1
        power
      else
        1.0
      end
    end

    def getUIversion
      WeaponToUI.new(self)
    end

    def to_s
      "Nombre: #{@name} - Tipo: #{type} - Usos #{uses}"
    end

    def clone
      Weapon.new(@name, @type.clone, @uses)
    end
  end
end
